tests/integration/ha_tests/helpers.py

```diff
@@ -0,0 +1,361 @@
+            timeout=TIMEOUT,
+        )
+
+        assert len(ops_test.model.applications[mongodb_application_name].units) == 3
+
+    return mongodb_application_name
+
+
+async def deploy_and_scale_application(ops_test: OpsTest) -> str:
+    """Deploys and scales the test application charm.
+
+    Args:
+        ops_test: The ops test framework
+    """
+    application_name = await get_application_name(ops_test, "application")
+
+    if application_name:
+        if len(ops_test.model.applications[application_name].units) != 1:
+            async with ops_test.fast_forward():
+                await scale_application(ops_test, application_name, 1)
+
+        return application_name
+
+    global application_charm
+    if not application_charm:
+        charm = await ops_test.build_charm("./tests/integration/ha_tests/application_charm/")
+        # Cache the built charm to avoid rebuilding it between tests
+        application_charm = charm
+
+    async with ops_test.fast_forward():
+        await ops_test.model.deploy(
+            application_charm,
+            application_name=APPLICATION_DEFAULT_APP_NAME,
+            num_units=1,
+        )
+
+        await ops_test.model.wait_for_idle(
+            apps=[APPLICATION_DEFAULT_APP_NAME],
+            status="waiting",
+            raise_on_blocked=True,
+            timeout=TIMEOUT,
+        )
+
+        assert len(ops_test.model.applications[APPLICATION_DEFAULT_APP_NAME].units) == 1
+
+    return APPLICATION_DEFAULT_APP_NAME
+
+
+def is_relation_joined(ops_test: OpsTest, endpoint_one: str, endpoint_two: str) -> bool:
+    """Check if a relation is joined.
+
+    Args:
+        ops_test: The ops test object passed into every test case
+        endpoint_one: The first endpoint of the relation
+        endpoint_two: The second endpoint of the relation
+    """
+    for rel in ops_test.model.relations:
+        endpoints = [endpoint.name for endpoint in rel.endpoints]
+        if endpoint_one in endpoints and endpoint_two in endpoints:
+            return True
+    return False
+
+
+async def get_process_pid(
+    ops_test: OpsTest, unit_name: str, container_name: str, process: str
+) -> int:
+    """Return the pid of a process running in a given unit.
+
+    Args:
+        ops_test: The ops test object passed into every test case
+        unit_name: The name of the unit
+        container_name: The name of the container in the unit
+        process: The process name to search for
+    Returns:
+        A integer for the process id
+    """
+    get_pid_commands = [
+        "ssh",
+        "--container",
+        container_name,
+        unit_name,
+        "pgrep",
+        process,
+    ]
+    return_code, pid, _ = await ops_test.juju(*get_pid_commands)
+
+    assert (
+        return_code == 0
+    ), f"Failed getting pid, unit={unit_name}, container={container_name}, process={process}"
+
+    stripped_pid = pid.strip()
+    if not stripped_pid:
+        return -1
```

Comment:
I think it would be better to raise an error here, AFAICT, this edge case of `-1` being returned isn't handled